go-controller/pkg/node/port_claim.go

```diff
@@ -120,61 +150,66 @@ func updateServicePortClaim(oldSvc, newSvc *kapi.Service) []error {
 	return errors
 }
 
-func (p *portClaimWatcher) open(port int32, protocol kapi.Protocol, svc *kapi.Service) error {
-	klog.V(5).Infof("Opening socket for service: %s/%s and port: %v", svc.Namespace, svc.Name, port)
-	var socket activeSocket
-	var socketError error
+func (p *portClaimWatcher) open(desc string, ip string, port int32, protocol kapi.Protocol, svc *kapi.Service) error {
+	klog.V(5).Infof("Opening socket for service: %s/%s, port: %v and protocol %s", svc.Namespace, svc.Name, port, protocol)
+
+	var localPort *utilnet.LocalPort
+	var portError error
 	switch protocol {
 	case kapi.ProtocolTCP:
-		listener, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
-		if err != nil {
-			socketError = err
-			break
-		}
-		socket = listener
+		localPort, portError = utilnet.NewLocalPort(desc, ip, "", int(port), utilnet.TCP)
```

Comment:
this has to receive the IP family, sees you are passing an empty string

```
// NewLocalPort returns a LocalPort instance and ensures IPFamily and IP are
// consistent and that the given protocol is valid.
func NewLocalPort(desc, ip string, ipFamily IPFamily, port int, protocol Protocol) (*LocalPort, error) {
```

https://github.com/kubernetes/utils/blob/420da100c033b61c6933e150195d023b39e44bbc/net/port.go#L66

Reply:
IP family is an optional parameter, if it is empty, and IP address is provided, LocalPort tries to resolve it according to the IP string, otherwise the local port bind to local addresses of any family.
https://github.com/kubernetes/utils/blob/420da100c033b61c6933e150195d023b39e44bbc/net/port.go#L55

Reply:
my fault, you are right